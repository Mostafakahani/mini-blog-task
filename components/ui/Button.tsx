import React from "react";

interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  variant?: "default" | "primary" | "secondary" | "danger" | "ghost";
  size?: "sm" | "md" | "lg";
  isLoading?: boolean;
}

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  (
    {
      className,
      children,
      variant = "default",
      size = "md",
      isLoading = false,
      disabled,
      ...props
    },
    ref
  ) => {
    const baseStyles =
      "w-full cursor-pointer font-medium rounded-lg !transition-all shadow-md hover:shadow-lg disabled:shadow-none disabled:cursor-not-allowed";

    const variants = {
      default: "bg-gray-100 hover:bg-gray-200 text-gray-900",
      primary:
        "bg-gradient-to-r from-blue-500 to-blue-600 text-white hover:from-blue-600 hover:to-blue-700 disabled:from-blue-400 disabled:to-blue-500",
      secondary: "bg-gray-600 hover:bg-gray-700 text-white",
      danger: "bg-red-600 hover:bg-red-700 text-white",
      ghost:
        "bg-transparent hover:bg-gray-100 text-gray-700 shadow-none hover:shadow-none",
    };

    const sizes = {
      sm: "px-3 py-1.5 text-sm",
      md: "px-4 py-2 text-base",
      lg: "px-6 py-3 text-lg",
    };

    return (
      <button
        ref={ref}
        disabled={disabled || isLoading}
        className={`${baseStyles} ${variants[variant]} ${sizes[size]} ${
          isLoading ? "opacity-70 cursor-not-allowed" : ""
        } ${disabled ? "opacity-50 cursor-not-allowed" : ""} ${className}`}
        {...props}
      >
        {isLoading ? (
          <div className="flex items-center gap-2">
            <span className="animate-spin">‚è≥</span>
            {children}
          </div>
        ) : (
          children
        )}
      </button>
    );
  }
);

Button.displayName = "Button";

export default Button;
