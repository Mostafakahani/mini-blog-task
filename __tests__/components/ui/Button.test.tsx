import Button from "@/components/ui/Button";
import { render, screen, fireEvent } from "@testing-library/react";
import React from "react";

describe("Button", () => {
  it("renders with children and proper base, variant, and size classes", () => {
    render(
      <Button variant="primary" size="lg">
        Primary Button
      </Button>
    );
    const buttonElement = screen.getByText("Primary Button").closest("button");
    expect(buttonElement).toBeInTheDocument();
    expect(buttonElement?.className).toMatch(
      /bg-gradient-to-r from-blue-500 to-blue-600/
    );
    expect(buttonElement?.className).toMatch(/px-6 py-3 text-lg/);
  });

  it("calls onClick when clicked", () => {
    const handleClick = jest.fn();
    render(<Button onClick={handleClick}>Click Me</Button>);
    const buttonElement = screen.getByText("Click Me").closest("button");
    fireEvent.click(buttonElement!); // ! ensures non-null element
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it("renders a loading state correctly", () => {
    render(<Button isLoading={true}>Loading</Button>);
    const buttonElement = screen.getByText("Loading").closest("button");
    expect(buttonElement).toBeDisabled();
    expect(buttonElement?.className).toMatch(/cursor-not-allowed/);
    // Check for loading spinner presence (⏳)
    expect(screen.getByText("⏳")).toBeInTheDocument();
  });

  it("renders a disabled state correctly", () => {
    render(<Button disabled={true}>Disabled</Button>);
    const buttonElement = screen.getByText("Disabled").closest("button");
    expect(buttonElement).toBeDisabled();
    expect(buttonElement?.className).toMatch(/cursor-not-allowed/);
  });

  it("applies additional props correctly", () => {
    render(
      <Button type="submit" aria-label="submit-button">
        Submit
      </Button>
    );
    const buttonElement = screen.getByText("Submit").closest("button");
    expect(buttonElement).toHaveAttribute("type", "submit");
    expect(buttonElement).toHaveAttribute("aria-label", "submit-button");
  });

  it("matches snapshot", () => {
    const { container } = render(
      <Button variant="secondary" size="sm">
        Snapshot Test
      </Button>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
