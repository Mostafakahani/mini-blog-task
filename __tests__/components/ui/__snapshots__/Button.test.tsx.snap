// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button matches snapshot 1`] = `
<button
  class="w-full cursor-pointer font-medium rounded-lg !transition-all shadow-md hover:shadow-lg disabled:shadow-none disabled:cursor-not-allowed bg-gray-600 hover:bg-gray-700 text-white px-3 py-1.5 text-sm    undefined"
>
  Snapshot Test
</button>
`;
